File: qabelbox/src/main/java/de/qabel/qabelbox/activities/BaseWizardActivity.java
package de.qabel.qabelbox.activities;

import android.app.FragmentTransaction;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import de.qabel.core.accounting.AccountingHTTP;
import de.qabel.qabelbox.R;
import de.qabel.qabelbox.fragments.BaseIdentityFragment;
import de.qabel.qabelbox.fragments.CreateIdentityHeaderFragment;
import de.qabel.qabelbox.helper.UIHelper;

/**
 * Created by danny on 11.01.2016.
 */
public abstract class BaseWizardActivity extends CrashReportingActivity {

    private String TAG = this.getClass().getSimpleName();

    public static final String FIRST_RUN = "first_run";

    public static final String P_IDENTITY = "identity_name";
    protected BaseWizardActivity mActivity;
    protected MenuItem mActionNext;
    private ActionBar actionBar;
    private CreateIdentityHeaderFragment mIdentityHeaderFragment;

    protected BaseIdentityFragment[] fragments;
    protected int step = 0;
    public int activityResult = RESULT_CANCELED;
    protected boolean mFirstRun;
    //values for create box account mode
    AccountingHTTP mAccounting;
    protected boolean canExit = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {

        super.onCreate(savedInstanceState);
        mActivity = this;
        mFirstRun = getIntent().getBooleanExtra(FIRST_RUN, true);
        overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out);
        setContentView(R.layout.activity_create_identity);
        setupToolbar();
        createFragments();
        actionBar.setTitle(getActionBarTitle());
    }

    private void setupToolbar() {

        Toolbar mToolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(mToolbar);
        actionBar = getSupportActionBar();
        assert getSupportActionBar() != null;
        actionBar.setDisplayShowHomeEnabled(true);
        actionBar.setDisplayHomeAsUpEnabled(true);
        actionBar.setHomeButtonEnabled(true);
        if (!canExit) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(false);
        }
        mToolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                onBackPressed();
            }
        });
    }

    private void createFragments() {

        mIdentityHeaderFragment = new CreateIdentityHeaderFragment();
        fragments = getFragmentList();
        FragmentTransaction ft = getFragmentManager().beginTransaction();
        ft.add(R.id.fragment_container_content, fragments[0]);
        ft.add(R.id.fragment_container_header, mIdentityHeaderFragment);
        ft.commit();
    }

    @Override
    public void onBackPressed() {

        int fragmentCount = getFragmentManager().getBackStackEntryCount();
        //check backstack if fragments exists
        if (step < fragments.length && fragmentCount > 0) {

            //check if last fragment displayed
            if (fragmentCount == fragments.length - 1) {
                //complete wizard
                activityResult = RESULT_OK;
                completeWizard();

                return;
            }
            //otherwise, popbackstack and update ui
            if (step > 0) {
                step--;
            }
            if (step == 0 && !canExit) {
                actionBar.setDisplayHomeAsUpEnabled(false);
                actionBar.setDisplayUseLogoEnabled(true);
            }
            getFragmentManager().popBackStack();
            mIdentityHeaderFragment.updateUI(getHeaderFragmentText());
            updateActionBar(step);
        } else {
            //return without finish the wizard
            if (canExit) {
                finish();
            } else {

                UIHelper.showDialogMessage(this, getString(R.string.dialog_headline_warning), String.format(getString(R.string.message_step_is_needed_or_close_app), getWizardEntityLabel()), R.string.yes, R.string.no, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {

                        finish();
                    }
                }, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {

                    }
                });
            }
        }
    }

    protected abstract String getHeaderFragmentText();

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {

        getMenuInflater().inflate(R.menu.ab_create_identity, menu);
        mActionNext = menu.findItem(R.id.action_next);
        updateActionBar(step);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        if (item.getItemId() == R.id.action_next) {
            handleNextClick();
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    public void handleNextClick() {

        String check = ((BaseIdentityFragment) getFragmentManager().findFragmentById(R.id.fragment_container_content)).check();
        //check if fragment ready to go to the next step
        if (check != null) {
            //no, show error message
            UIHelper.showDialogMessage(this, R.string.dialog_headline_info, check);
        } else {

            //check if currently last step
            if (step == fragments.length - 1) {
                activityResult = RESULT_OK;
                completeWizard();
                return;
            }
            //no... go to next step

            if (canShowNext(step)) {
                showNextFragment();
            }
        }
    }

    /**
     * override this if you need special handling on next click
     *
     * @param step
     * @return
     */
    protected boolean canShowNext(int step) {
        return true;
    }

    protected void showNextFragment() {

        step++;
        if (step == fragments.length - 1) {
            canExit = true;
        }
        actionBar.setDisplayHomeAsUpEnabled(true);
        actionBar.setDisplayUseLogoEnabled(false);

        mIdentityHeaderFragment.updateUI(getHeaderFragmentText());
        getFragmentManager().beginTransaction().replace(R.id.fragment_container_content, fragments[step]).addToBackStack(null).commit();
        updateActionBar(step);
    }

    /**
     * refresh actionbar depends from current wizard state
     *
     * @param step current step number
     */
    protected void updateActionBar(int step) {

        //update icons
        if (step == 0) {
            mActionNext.setVisible(false);
        } else if (step < fragments.length - 1) {
            mActionNext.setVisible(true);
            actionBar.setDisplayShowHomeEnabled(true);
            actionBar.setHomeButtonEnabled(true);
            mActionNext.setTitle(R.string.next);
        } else {
            mActionNext.setVisible(true);
            actionBar.setDisplayHomeAsUpEnabled(false);
            actionBar.setDisplayShowTitleEnabled(false);
            mActionNext.setTitle(R.string.finish);
        }

        //update subtitle
        if (step == 0) {
            actionBar.setSubtitle(null);
        } else if (step < fragments.length - 1) {
            actionBar.setSubtitle(getString(R.string.step_x_from_y).replace("$1", step + "").replace("$2", (fragments.length - 2) + ""));
        }
    }

    protected abstract int getActionBarTitle();

    protected abstract BaseIdentityFragment[] getFragmentList();

    public abstract void completeWizard();

    public interface NextChecker {

        String check(View view);
    }

    protected abstract String getWizardEntityLabel();
}

File: qabelbox/src/main/java/de/qabel/qabelbox/activities/BaseWizardActivity.kt
package de.qabel.qabelbox.activities

import android.app.Activity
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import de.qabel.qabelbox.R
import de.qabel.qabelbox.fragments.BaseIdentityFragment
import de.qabel.qabelbox.fragments.CreateIdentityHeaderFragment
import de.qabel.qabelbox.helper.Formatter
import de.qabel.qabelbox.helper.UIHelper
import kotlinx.android.synthetic.main.activity_create_identity.*

abstract class BaseWizardActivity : CrashReportingActivity() {

    protected abstract val fragments: Array<BaseIdentityFragment>

    protected abstract val actionBarTitle: Int
    protected abstract val wizardEntityLabel: String

    protected abstract val headerFragmentText: String
    protected open val headerSecondLine: String = ""
    protected open val headerThirdLine: String = ""

    internal var nextAction: MenuItem? = null
    lateinit var headerFragment: CreateIdentityHeaderFragment

    protected var step = 0
    var activityResult = Activity.RESULT_CANCELED

    //Indicated that the wizard is cancelable
    protected var canExit = false

    interface NextChecker {
        fun check(view: View): String?
    }

    abstract fun completeWizard()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out)
        setContentView(R.layout.activity_create_identity)
        createFragments()
    }

    override fun onResume() {
        super.onResume()
        setupToolbar()
        toolbar.setTitle(actionBarTitle)
        if (!canExit) {
            supportActionBar?.setDisplayHomeAsUpEnabled(false)
        }
    }

    private fun setupToolbar() {
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayShowHomeEnabled(true)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setHomeButtonEnabled(true)

        toolbar.setNavigationOnClickListener { onBackPressed() }
    }

    private fun createFragments() {
        headerFragment = CreateIdentityHeaderFragment()
        val ft = fragmentManager.beginTransaction()
        ft.add(R.id.fragment_container_content, fragments[0])
        ft.add(R.id.fragment_container_header, headerFragment)
        ft.commit()
    }

    protected fun checkEMailAddress(editText: String): String? {
        val error = editText.length < 1
        if (error) {
            return getString(R.string.create_identity_enter_all_data)
        }
        if (!Formatter.isEMailValid(editText)) {
            return getString(R.string.email_address_invalid)
        }
        return null
    }

    override fun onBackPressed() {

        val fragmentCount = fragmentManager.backStackEntryCount
        //check backstack if fragments exists
        if (step < fragments.size && fragmentCount > 0) {

            //check if last fragment displayed
            if (fragmentCount == fragments.size - 1) {
                //complete wizard
                activityResult = Activity.RESULT_OK
                completeWizard()

                return
            }
            //otherwise, popbackstack and update ui
            if (step > 0) {
                step--
            }
            if (step == 0 && !canExit) {
                supportActionBar?.setDisplayHomeAsUpEnabled(false)
                supportActionBar?.setDisplayUseLogoEnabled(true)
            }
            fragmentManager.popBackStack()
            headerFragment.updateUI(headerFragmentText)
            updateActionBar(step)
        } else {
            //return without finish the wizard
            if (canExit) {
                finish()
            } else {
                UIHelper.showDialogMessage(this, getString(R.string.dialog_headline_warning), String.format(getString(R.string.message_step_is_needed_or_close_app), wizardEntityLabel), R.string.yes, R.string.no, { dialog, which -> finish() }) { dialog, which -> }
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.ab_create_identity, menu)
        nextAction = menu.findItem(R.id.action_next) as MenuItem
        updateActionBar(step)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {

        if (item.itemId == R.id.action_next) {
            handleNextClick()
            return true
        }

        return super.onOptionsItemSelected(item)
    }

    open fun handleNextClick() {

        val check = (fragmentManager.findFragmentById(R.id.fragment_container_content) as BaseIdentityFragment).check()
        //check if fragment ready to go to the next step
        if (check != null) {
            //no, show error message
            UIHelper.showDialogMessage(this, R.string.dialog_headline_info, check)
        } else {

            //check if currently last step
            if (step == fragments.size - 1) {
                activityResult = Activity.RESULT_OK
                completeWizard()
                return
            }
            //no... go to next step

            if (canShowNext(step)) {
                showNextFragment()
            }
        }
    }

    /**
     * override this if you need special handling on next click

     * @param step
     * *
     * @return
     */
    protected open fun canShowNext(step: Int): Boolean {
        return true
    }

    protected fun showNextFragment() {
        step++
        if (step == fragments.size - 1) {
            canExit = true
        }
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setDisplayUseLogoEnabled(false)

        headerFragment.updateUI(headerFragmentText, headerSecondLine, headerThirdLine)
        fragmentManager.beginTransaction().replace(R.id.fragment_container_content, fragments[step]).addToBackStack(null).commit()
        updateActionBar(step)
    }

    /**
     * refresh actionbar depends from current wizard state

     * @param step current step number
     */
    protected open fun updateActionBar(step: Int) {

        //update icons
        if (step == 0) {
            nextAction?.isVisible = false
        } else if (step < fragments.size - 1) {
            nextAction?.isVisible = true
            supportActionBar?.setDisplayShowHomeEnabled(true)
            supportActionBar?.setHomeButtonEnabled(true)
            nextAction?.setTitle(R.string.next)
        } else {
            nextAction?.isVisible = true
            supportActionBar?.setDisplayHomeAsUpEnabled(false)
            supportActionBar?.setDisplayShowTitleEnabled(false)
            nextAction?.setTitle(R.string.finish)
        }

        //update subtitle
        if (step == 0) {
            supportActionBar?.subtitle = null
        } else if (step < fragments.size - 1) {
            actionBar?.subtitle = getString(R.string.step_x_from_y).
                    replace("$1", step.toString()).
                    //First and last excluded -> -2
                    replace("$2", (fragments.size - 2).toString())
        }
    }

}
